api/services/sandbox/v1/sandbox.proto

```diff
@@ -190,3 +191,14 @@ message ControllerMetricsRequest {
 message ControllerMetricsResponse {
 	types.Metric metrics = 1;
 }
+
+message ControllerUpdateResourceRequest {
+	string sandbox_id = 1;
+	string sandboxer = 2;
+	containerd.types.ResourceOp op = 3;
+	containerd.types.TaskResource resource = 4;
```

Comment:
Can you add some clarity, how this intended to work?

`TaskResource` is pretty similar to [`CreateTaskRequest`](https://github.com/containerd/containerd/blob/87bab6cdc7767e30681d909fec0570647550c52d/api/runtime/task/v3/shim.proto#L53).
If I call this with `op=ADD`, how is this different from creating or execing a task?

Reply:
If we assume that the task server may run inside the sandbox, for example, in a VM, then the task server can not directly append the resources on host to the VM as it is running inside the VM. So maybe we can call the UpdateResource API to controller to append the resource to the VM first.

Reply:
This API is just for sandbox controller to append or remove resources related to the task from the sandbox, for example, for VM based sandbox, we need to hotplug some devices or add some cpus or memories to the VM before we start a new task in the VM.

Reply:
Oh, ok, I'm probably a bit confused with selection of fields for `TaskResource` struct.
It looks more suitable for launching processes to me. How TaskResource fields would look like for hot plug example?

Reply:
Yes the fields in `TaskResource` looks like the information for starting a task or execing a process, This is actually because we call `UpdateResource` of a sandbox before or after task or process start or delete. So we put the all the information related to the Task or Process into the `TaskResource`, including the spec, the mounts, and the stdios, For a VM based sandbox, we may have to get the sources from the mounts, and hotplug a block device into VM, or pass a directory into VM through virtiofs, we may also get the rootfs from spec, if it is devicemapper, we also need to hotplug the logical block device of rootfs into VM. We are not sure how a sandbox controller handle the resources related to the task, so we put all the information of a task or a process into the `TaskResource` struct, the information is complete and different kinds of sandbox controllers can choose what to do to the sandbox based on the complete information.

Reply:
Given the diverse nature of a "resource", would it make sense to keep task resource abstract? So we won't be limited by set of fields which are suitable for launching tasks, but not very well suitable for other kind of resources. For example:

```proto
message ControllerUpdateResourceRequest {
	string sandbox_id = 1;
	string sandboxer = 2;
        // Uniquely identifies resource to be added/updated/deleted.
        string resource_id = 1;
        // Type of operation (added/update/delete).
	containerd.types.ResourceOp op = 3;
        // Resource data.
	google.protobuf.Any resource = 4;
}
```

This way we can define a list of resource updaters (and it'll be possible to add new types in future).

Reply:
done, please take a look.